snuba/datasets/plans/translator/visitor.py

```diff
@@ -0,0 +1,158 @@
+            subscriptables=[*self.subscriptables, *spec.subscriptables],
+            functions=[*self.functions, *spec.functions],
+            curried_functions=[*self.curried_functions, *spec.curried_functions],
+            arguments=[*self.arguments, *spec.arguments],
+            lambdas=[*self.lambdas, *spec.lambdas],
+        )
+
+
+TExpIn = TypeVar("TExpIn", bound=Expression)
+
+
+class ExpressionMapper(ABC, Generic[TExpIn, TExpOut]):
+    """
+    One translation rule used by the MappingExpressionTranslator to translate an
+    expression.
+
+    see MappingExpressionTranslator for more context
+    """
+
+    @abstractmethod
+    def attemptMap(
+        self, expression: TExpIn, children_translator: ExpressionVisitor[TExpOut]
+    ) -> Optional[TExpOut]:
+        """
+        Translates an expression if this rule matches such expression. If not, it
+        returns None.
+
+        It receives an instance of a translating visitor to take care of children
+        that this rule is not capable, or not supposed to, translate. These node
+        should be delegated to the children_translator.
+        """
+        raise NotImplementedError
+
+
+class MappingExpressionTranslator(ExpressionVisitor[TExpOut]):
+    """
+    Translates a Snuba query expression into an physical query expression (like a
+    Clickhouse query expression).
+
+    The translation of every node in the expression is performed by a series of rules
+    that extend ExpressionMapper.
+    Rules are applied in sequence. Given an expression, the first valid rule for such
+    expression is applied and the result is returned. If no rule can translate such
+    expression an exception is raised.
+    A rule can delegate the translation of its children back to this visitor.
+
+    Each rule only has context around the expression provided and its children. It does
+    not have general context around the query or around the expression's ancestors in
+    the AST.
+    This approach implies that, while rules are specific to the relationship between
+    dataset (later entities) and storage, this class keeps the responsibility of
+    orchestrating the translation process.
+
+    It is possible to compose different, independently defined, sets of rules that are
+    applied in a single pass over the AST.
+    This allows us to support joins and multi-step translations (for multi table
+    storages) as an example:
+    Joins can be supported by simply concatenating rule sets associated with each storage.
+    Multi-step (still TODO) translations can be supported by applying a second sequence of
+    rules to the result of the first one for each node in the expression to be translated.
+
+    The visitor approach, while a bit verbose, allows us to impose some static
+    guarantees of correctness of the translator. It is impossible, for example, to add a
+    new node type to the Snuba AST and ignoring it in the translator (a method
+    implementation would be missing).
```

Comment:
To me the advantage of building this with a visitor is that mapping rules (ExpressionMapper) can be Generic with the expression type as parameter. This mean we do not need to check the expression type with an `isinstance` call at the very beginning of each mapper.